### ISSUE TITLE
Deadlock in crash recovery

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#1051]
- **Version:** Git Master Branch
- **OS:** Windows

I'm running a collaboration server on a dual core machine. In this case the-machinery only spawns one "task thread".

When saving the project, this spawns a task. I got a deadlock of this task with the callstack:
```
 	the-machinery.exe!is_task_done(unsigned __int64 id) Line 168	C
>	the-machinery.exe!stop_recording(tm_crash_recovery_o * cr) Line 120	C
 	the-machinery.exe!start_recording(tm_crash_recovery_o * cr, const char * project, tm_the_truth_o * tt, tm_tt_id_t asset_root) Line 90	C
 	[Inline Frame] the-machinery-dll.tm_hot_reload.dll!save_to_directory(tm_application_o * dir, const char *) Line 5167	C
 	the-machinery-dll.tm_hot_reload.dll!private__save_all_async(void * data, unsigned __int64 task_id) Line 5197	C
 	the-machinery.exe!process_one_task() Line 77	C
```

https://github.com/OurMachinery/themachinery/blob/9dee2573507d209f2e53ee6ce7a4af897171c49c/foundation/crash_recovery.c#L119-L122

The task which we wait for is the second of three unstarted tasks in the task system:
```
[0]	{f=tm_creation_graph.tm_hot_reload.dll!0x00007ffe28811d90 data=0x000002be34e90680 id=430819 ...}	task_t
[1]	{f=0x00007ff66a282200 {the-machinery.exe!write_file_task(void *, unsigned __int64)} data=0x0000025d99a03520 ...}	task_t
[2]	{f=0x00007ffe28981e90 {the-machinery-dll.tm_hot_reload.dll!save_task(void *, unsigned __int64)} data=...}	task_t
```

Since there only is one task thread, we will spinlock forever waiting for the unstarted task to be done.

I guess using `is_task_done_else_assist()` here would solve my issue.

The reason why I created this issue instead of just doing a PR with the fix is that I wonder how we could systematically solve issues like this.

Personal reflections:
I don't think there should be any busywaiting using `is_task_done()` at all. That function is only valuable for checking if the result from the task is ready to be used right now, otherwise we should bail out, either checking again next frame, or we can spawn a new task which will run later and pick up the work, or we can call `is_task_done_else_assist()`. But never spinlock style waiting for a task. And this should at least be part of documentation of `is_task_done()`. The task system could also be kind and provide a function like "wait_for_task()" which tries to assist, and if there is no tasks to pick then it blocks on a semaphore waiting on any task to complete so we never spin for a long time. Or at least yield. Having a standardized good default for waiting on a task has proven useful to me before, especially when different platforms should do different things, or when you want to do different things depending on which thread calls it. Especially to avoid code like this:
https://github.com/OurMachinery/themachinery/blob/bd8e96b49848c5a2fc8bb6e0879b2340784be5dd/the_machinery/the_machinery.c#L5906-L5907

I guess there are many cases worth thinking about, and I don't have all of them in my head. But here goes a few thoughts:
* When there is only one task thread (or at least very few compared to the number of tasks): In this case any blocking wait has a big risk of causing a deadlock since the thread will now not be usable for any other work. This is what the current github issue suffers from.
* When the number of task threads are more than then amount of work: Is it ever valuable for a task to spinlock wait on other tasks? In my mind the tasks are allowed to run for a long time, so a spinlock might cause a busywait using 100% CPU on that thread lasting many milliseconds. A real lock (mutex, semaphore, critical section) would let the CPU rest or let it do other work while waiting on the other task which would be a better user experience (and save the environment, yay! ;) )
* When a fiber want to wait for a task: This should generally never happen. Code run by the job system should always just try again the next frame, never stall on work done by the task system. But I myself have made mistakes here, in code for startup, shutdown and modal dialogs where I try to cancel a task and then busywait for it to happen and it took forever.
* Library code which wants to wait on a task but is generic and it can't know if the calling code is in a job or a task. In such a case the task system could know how to wait based on which thread the call is made from. This knowledge is hard to have on the calling side.

**Comment by niklas-ourmachinery** [comment]
I'm inclined to agree.

There is a potential issue with using `is_task_done_else_assist()` everywhere.

Suppose there are two tasks A and B... A takes 0.1s to complete and B takes 100s to complete.

We need to wait for A so we do `while (!is_task_done_else_assist(A))`. Another thread has already started work on A but is not complete yet. In this case we will pick up B and our function will have to wait for 100s when really it only needed to wait for 0.1s.

But I'm not sure we need to address this, since the task list is a queue with limited bandwidth anyway. I.e. it might be that five 100s tasks (B_1, B_2, B_3, B_4, B_5) got scheduled before A and in this case we would have to wait 100 s anyway.

So I think for now we should just make a `wait_for_task()` function that calls `is_task_done_else_assist()`. If we run into issues similar to the A/B situation above we can maybe try to fix it with making the `wait_for_task()` function a bit smarter with which things it "assists" with, though that can get tricky too. (For example, what if we decide *not* to assist with B, in the case above, in order to not wait longer than necessary, but it turns out that B was actually a subtask spawned by A. Now we're risking deadlock again.)

I'll put together a PR and have @karl-zylinski-ourmachinery take a peek at it before merging.

**Comment by niklas-ourmachinery** [comment]
PR here: https://github.com/OurMachinery/themachinery/pull/889

**Comment by niklas-ourmachinery** [comment]
Merged the PR.
